### PR TITLE
Peltool: Performance improvement

### PR DESCRIPTION
**Pull request by harsh-agarwal1**
Enhances performance of -a option by printing as soon as each PEL 
is parsed. 
This change minimizes the overhead associated with storing and 
processing large strings, resulting in faster output generation.

Tested Non-BMC env:
```bash
time python3 -m pel.peltool.peltool -a -p 3kpels/
1m3.707s    ->  0m38.749s
```

Optimize performance by implementing caching mechanisms for
component IDs, user data parsers, callout parsers, and SRC
modules.
These changes reduce redundant module imports by storing previously
loaded modules in dictionaries, streamlining the parsing process
and enhancing overall efficiency.

Tested:
```bash
Non-BMC env:
-n: 0.941s    ->  0.283s
-l: 6.100s    ->  4.225s
-a: 52.740s   ->  17.371s

BMC env:
-n: 4.287s   ->  4.217s
-l: 1m24.911s ->  14.125s
-a: 1m37.431s ->  24.958s
```

Signed-off-by: Harsh Agarwal Harsh.Agarwal@ibm.com

### CODE REVIEW

**Comment by spinler** [comment]
When I used the -a and redirected it to a file, when it tried to parse that file the error showed up within the JSON output:

```
            "00000650     46737043  6B73756D  00000000  A13DDA96     FspCksum.....=..",
            "00000660     4F757243  6B73756D  00000000  00000000     OurCksum........"
        ]
    }
}Failed to parse Private Header, section ID = 5b0a

]
```

And also with a garbage file named 1_1 in that dir it shows up in the middle:
```
            "Jan 04 19:40:31 systemd[1]: phosphor-regulators.service: Sent signal SIGHUP to main process 483 (phosphor-regula) on client request."
        ]
    }
}Exception: No PEL parsed for 1_1: range check failure
,
{
    "Private Header": {
```

I think probably printing those to stderr would at least get them out of the JSON?

**Comment by spinler** [comment]
Also here I think the commit message can just say what the fix is, and then in the body it can say it's for performance reasons.

**Comment by harsh-agarwal1** [comment]
Addressed [comment](https://github.com/ibm-openbmc/openpower-pel-parsers/pull/69#issuecomment-2399975352) in #70 .

**Review by RameshIyyar — APPROVED** [review]

File: modules/pel/peltool/peltool.py [hunk 1]
```diff
@@ -625,7 +625,8 @@ def extractAllPELsData(path: str, config: Config):
     Prints a JSON-formatted string containing PEL data from all valid files.
     """
     root, file_list = getFileList(path, config.extension, config.rev)
-    allPELsData = "[ \n"
+    print("[")
```
Comment (spinler) [inline modules/pel/peltool/peltool.py:628]: This was wrong before, but this [ is only needed if not config.hex

File: modules/pel/peltool/peltool.py [hunk 2]
```diff
@@ -634,14 +636,18 @@ def extractAllPELsData(path: str, config: Config):
                 _, json_string = parsePEL(stream, config, False)
                 if json_string:
                     if not config.hex:
-                        allPELsData = allPELsData + json_string + ",\n"
+                        if firstPELPrinted:
+                            print(",")
+                        print(json_string, end = "", flush=True)
```
Comment (RameshIyyar) [inline modules/pel/peltool/peltool.py:641]: Why are you flushing? I think you mentioned that `flush` does not make any improvement while printing into terminal, right?

File: modules/pel/peltool/peltool.py [hunk 3]
```diff
@@ -634,14 +636,18 @@ def extractAllPELsData(path: str, config: Config):
...
+                        firstPELPrinted = True
                     else:
                         printPELInHexFormat(data)
             except Exception as e:
                 print(f"Exception: No PEL parsed for {file}: {e}", file=sys.stderr)
     if not config.hex:
-        allPELsData = allPELsData[:-2] + "\n]"
-        print(allPELsData)
+        if firstPELPrinted:
+            print()
```
Comment (RameshIyyar) [inline modules/pel/peltool/peltool.py:649]: Why new line?

Comment (harsh-agarwal1) [reply modules/pel/peltool/peltool.py:649]: To match with cpp output

**Review by RameshIyyar — APPROVED** [review]

**Comment by spinler** [comment]
I got this error:
```
Exception: No PEL parsed for 2023051612144611_50003F01: name 'calloutParserMod' is not defined
```

File: modules/pel/peltool/comp_id.py [hunk 4]
```diff
@@ -1,27 +1,37 @@
...
+pelConfigRootPath = "/usr/share/phosphor-logging/pels"
+attemptedToParseCompIDs = False
+
+def getAllCreatorsCompIDs():
+    global attemptedToParseCompIDs
+    if attemptedToParseCompIDs:
+        return
+
+    attemptedToParseCompIDs = True
+    componetsConfigPath = ""
```
Comment (RameshIyyar) [inline modules/pel/peltool/comp_id.py:17]: > componetsConfigPath

typo `components`

File: modules/pel/peltool/src.py [hunk 5]
```diff
@@ -295,10 +306,18 @@ def parse(self, hexwords: list) -> str:
...
+            if srcParserMod in srcParsers:
+                cls = srcParsers[srcParserMod]
+                if cls is None:
+                    # The module, which was previously checked, is not found.
+                    return ""
+            else:
+                cls = importlib.import_module(srcParserMod)
+                srcParsers[srcParserMod] = cls
         except:
+            calloutParsers[srcParserMod] = None
```
Comment (RameshIyyar) [inline modules/pel/peltool/src.py:320]: > calloutParsers[srcParserMod]

Its wrong, I think it should be `srcParsers[srcParserMod]`

**Review by RameshIyyar — APPROVED** [review]

**Review by spinler — APPROVED** [review]

**Pull request merged**